block/submit.go

```diff
@@ -2,200 +2,210 @@ package block
-			total -= m.Conf.BlockBatchMaxSizeBytes
+	})
 
+	eg.Go(func() error {
+		// 'submitter': this thread actually creates and submits batches
+		timeLastSubmission := time.Now()
+		for {
 			select {
 			case <-ctx.Done():
-				return
-			case toSubmit <- struct{}{}:
-				m.logger.Info("Enough bytes to build a batch have been accumulated. Sent signal to submit the batch.")
-			default:
-				m.logger.Error("Enough bytes to build a batch have been accumulated, but too many batches are pending submission.	 " +
-					"Pausing block production until a batch submission signal is consumed.")
-
-				evt := &events.DataHealthStatus{Error: fmt.Errorf("submission channel is full: %w", gerrc.ErrResourceExhausted)}
-				uevent.MustPublish(ctx, m.Pubsub, evt, events.HealthStatusList)
-
-				/*
-					Now we block until earlier batches have been submitted. This has the effect of not consuming the producedSizeC,
-					which will stop new block production.
-				*/
-				select {
-				case <-ctx.Done():
-					return
-				case toSubmit <- struct{}{}:
+				return ctx.Err()
+			case <-submitter.C:
+			}
+			pending := pendingBytes.Load()
+			// while there are accumulated blocks, create and submit batches!!
+			for ctx.Err() == nil && (0 < pending && (maxBatchTime < time.Since(timeLastSubmission)) || maxBatchBytes < pending) {
+				nConsumed, err := createAndSubmitBatchGetSizeEstimate(min(pending, maxBatchBytes))
+				if err != nil {
+					return fmt.Errorf("create and submit batch: %w", err)
 				}
-
-				evt = &events.DataHealthStatus{Error: nil}
-				uevent.MustPublish(ctx, m.Pubsub, evt, events.HealthStatusList)
-
-				m.logger.Info("Resumed block production.")
+				timeLastSubmission = time.Now()
+				pending = pendingBytes.Add(^(nConsumed - 1)) // subtract
 			}
+			counter.Nudge()
 		}
-		select {
-		case <-ctx.Done():
-			return
-		case size := <-m.producedSizeC:
-			total += size
-		}
-	}
-}
-
-// HandleSubmissionTrigger processes the sublayer submission trigger event. It checks if there are new blocks produced since the last submission.
-// If there are, it attempts to submit a batch of blocks. It then attempts to produce an empty block to ensure IBC messages
-// pass through during the batch submission process due to proofs requires for ibc messages only exist on the next block.
-// Finally, it submits the next batch of blocks and updates the sync target to the height of the last block in the submitted batch.
-func (m *Manager) HandleSubmissionTrigger() error {
-	// Load current sync target and height to determine if new blocks are available for submission.
+	})
 
-	startHeight := m.NextHeightToSubmit()
-	endHeightInclusive := m.State.Height()
+	return eg.Wait()
+}
 
-	if endHeightInclusive < startHeight {
-		return nil // No new blocks have been produced
+func (m *Manager) CreateAndSubmitBatchGetSizeEstimate(maxSize uint64) (uint64, error) {
```

Comment:
`GetSizeEstimate` in the func name is redundent.
u can name the `(uint64, error)` if for explicitily if u wish
godoc

Reply:
I followed omris suggestion instead, but good point